services/rsources/handler.go

```diff
@@ -106,24 +106,107 @@ func (*sourcesHandler) IncrementStats(ctx context.Context, tx *sql.Tx, jobRunId
 	return err
 }
 
-func (*sourcesHandler) AddFailedRecords(_ context.Context, _ *sql.Tx, _ string, _ JobTargetKey, _ []json.RawMessage) error {
-	return nil
+func (*sourcesHandler) AddFailedRecords(ctx context.Context, tx *sql.Tx, jobRunId string, key JobTargetKey, records []json.RawMessage) (err error) {
+	stmt, err := tx.Prepare(`insert into "rsources_failed_keys" (
```

Comment:
I am wondering if an insert with multiple rows would be a better choice here, both in terms of performance and consistency (no partial failures).

Depending on the domain partial failures might not be appropriate

Reply:
`pq` doesn't appear to have support for multi-row insert, plus we will also need to cater for the limit of 1000 rows per multi-row insert...
That being said, I don't believe it is worth the trouble